lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+            if (!annotation.isClass(
+                    "sifive.enterprise.grandcentral.ViewAnnotation"))
+              return false;
+            auto tpe = annotation.getMember<StringAttr>("type"),
+                 name = annotation.getMember<StringAttr>("name");
+            auto id = annotation.getMember<IntegerAttr>("id");
+            if (!tpe) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain a 'type' field with a 'StringAttr' value";
+              goto FModuleOp_error;
+            }
+            if (!id) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain an 'id' field with an 'IntegerAttr' value";
+              goto FModuleOp_error;
+            }
+            if (!name) {
+              op.emitOpError()
+                  << "has a malformed "
+                     "'sifive.enterprise.grandcentral.ViewAnnotation' that did "
+                     "not contain a 'name' field with a 'StringAttr' value";
+              goto FModuleOp_error;
+            }
+
+            // If this is a companion, then:
+            //   1. Insert it into the companion map
+            //   2. Create a new mapping module.
+            //   3. Instatiate the mapping module in the companion.
+            //   4. Check that the companion is instantated exactly once.
+            //   5. Set attributes on that lone instance so it will become a
+            //      bind if extraction information was provided.
+            if (tpe.getValue() == "companion") {
+              builder.setInsertionPointToEnd(circuitOp.getBody());
+
+              // Create the mapping module.
+              auto mappingName =
+                  getNamespace().newName(name.getValue() + "_mapping");
+              auto mapping = builder.create<FModuleOp>(
+                  circuitOp.getLoc(), builder.getStringAttr(mappingName),
+                  SmallVector<ModulePortInfo>({}));
+              auto *ctx = builder.getContext();
+              mapping->setAttr(
+                  "output_file",
+                  hw::OutputFileAttr::get(
+                      getOutputDirectory(),
+                      StringAttr::get(ctx, mapping.getName() + ".sv"),
+                      BoolAttr::get(ctx, true), BoolAttr::get(ctx, true), ctx));
+              companionIDMap.insert({id, {name.getValue(), op, mapping}});
+
+              // Instantiate the mapping module inside the companion.
+              builder.setInsertionPointToEnd(op.getBodyBlock());
+              builder.create<InstanceOp>(circuitOp.getLoc(),
+                                         SmallVector<Type>({}),
+                                         mapping.getName(), mapping.getName());
```

Comment:
Passing a StringRef into the builder causes it to look up the attr again.  You could save the StringAttr from when you create `mapping` above and reuse it here.  You could also save `BoolAttr(ctxt, true)`  which is reused all over the place.

Reply:
Using the attrs is proving tedious. I'll leave this as is for readability. Switching to a locally cached `BoolAttr` sounds great.